README
	oauth-utils - command line utilities for oauth

  Copyright (C) 2008,2009 Robin Gareus

  oauth_utils is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

  see http://rg42.org/oss/oauth/ for examples and usage info

Comments are welcome.

	- Robin Gareus <robin@gareus.org>


ChangeLog
2010-06-01  Robin Gareus  <robin@gareus.org>
	* version 0.3.3
	* updated to liboauth 0.8.6 API

2009-08-17  Robin Gareus  <robin@gareus.org>
	* version 0.3.1
	* minor fixes ('-p' parameter)

2009-07-31  Robin Gareus  <robin@gareus.org>
	* version 0.3.0
	* added support for GET/PUT/HEAD request (Stephen Paul Weber)
	* commandline options for OAuth Code 1.0 Rev A

2008-10-19  Robin Gareus  <robin@gareus.org>
	* version 0.2.3
	* added oauthverify
	* updated man-pages
	* reworked output (stdout/stderr)
	* enforce required oauth parameters

2008-10-17  Robin Gareus  <robin@gareus.org>
	* version 0.2.0
	* escape parameters when saving keyfile
	* added curl commandline output
	* reading SHA1-key from consumer-secret :(

2008-10-16  Robin Gareus  <robin@gareus.org>
	* version 0.1.5
	* fixed segfault on filename-check.

2008-10-16  Robin Gareus  <robin@gareus.org>
	* version 0.1.4
	* updated debian packaging system

2008-10-16  Robin Gareus  <robin@gareus.org>
	* version 0.1.3
	* renamed oauth_utils to oauth-utils
	* debianized it
	* fixed typos, built-env and updated example.sh
	* changed parser to allow than two return parameters (xoauth_)

2008-10-16  Robin Gareus  <robin@gareus.org>
	* version 0.1.0
	* oauthsign works and is mostly documented.

2008-10-13  Robin Gareus  <robin@gareus.org>

	* oauth_utils: initial version.



src/system.h
/* system-dependent definitions for oauth-utils

   Written for GNU fileutils.
   Adapted for autoproject by James R. Van Zandt.

   Copyright (C) 89, 91, 92, 93, 94, 1996 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#if HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

/* Include sys/types.h before this file.  */

#include <sys/stat.h>

#ifdef STAT_MACROS_BROKEN
#undef S_ISBLK
#undef S_ISCHR
#undef S_ISDIR
#undef S_ISFIFO
#undef S_ISLNK
#undef S_ISMPB
#undef S_ISMPC
#undef S_ISNWK
#undef S_ISREG
#undef S_ISSOCK
#endif /* STAT_MACROS_BROKEN.  */

#ifndef S_IFMT
#define S_IFMT 0170000
#endif
#if !defined(S_ISBLK) && defined(S_IFBLK)
#define S_ISBLK(m) (((m) & S_IFMT) == S_IFBLK)
#endif
#if !defined(S_ISCHR) && defined(S_IFCHR)
#define S_ISCHR(m) (((m) & S_IFMT) == S_IFCHR)
#endif
#if !defined(S_ISDIR) && defined(S_IFDIR)
#define S_ISDIR(m) (((m) & S_IFMT) == S_IFDIR)
#endif
#if !defined(S_ISREG) && defined(S_IFREG)
#define S_ISREG(m) (((m) & S_IFMT) == S_IFREG)
#endif
#if !defined(S_ISFIFO) && defined(S_IFIFO)
#define S_ISFIFO(m) (((m) & S_IFMT) == S_IFIFO)
#endif
#if !defined(S_ISLNK) && defined(S_IFLNK)
#define S_ISLNK(m) (((m) & S_IFMT) == S_IFLNK)
#endif
#if !defined(S_ISSOCK) && defined(S_IFSOCK)
#define S_ISSOCK(m) (((m) & S_IFMT) == S_IFSOCK)
#endif
#if !defined(S_ISMPB) && defined(S_IFMPB) /* V7 */
#define S_ISMPB(m) (((m) & S_IFMT) == S_IFMPB)
#define S_ISMPC(m) (((m) & S_IFMT) == S_IFMPC)
#endif
#if !defined(S_ISNWK) && defined(S_IFNWK) /* HP/UX */
#define S_ISNWK(m) (((m) & S_IFMT) == S_IFNWK)
#endif

#ifndef S_IEXEC
#define S_IEXEC S_IXUSR
#endif

#ifndef S_IXUSR
#define S_IXUSR S_IEXEC
#endif
#ifndef S_IXGRP
#define S_IXGRP (S_IEXEC >> 3)
#endif
#ifndef S_IXOTH
#define S_IXOTH (S_IEXEC >> 6)
#endif
#ifndef S_IXUGO
#define S_IXUGO (S_IXUSR | S_IXGRP | S_IXOTH)
#endif

#if !defined(HAVE_MKFIFO)
#define mkfifo(path, mode) (mknod ((path), (mode) | S_IFIFO, 0))
#endif

#ifdef HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif

/* <unistd.h> should be included before any preprocessor test
   of _POSIX_VERSION.  */
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifndef STDIN_FILENO
#define STDIN_FILENO 0
#endif

#ifndef STDOUT_FILENO
#define STDOUT_FILENO 1
#endif

#ifndef STDERR_FILENO
#define STDERR_FILENO 2
#endif

/* FIXME: Don't use _POSIX_VERSION.  */
#ifndef _POSIX_VERSION
off_t lseek ();
#endif

#ifdef TM_IN_SYS_TIME
#include <sys/time.h>
#else
#include <time.h>
#endif

/* Since major is a function on SVR4, we can't use `ifndef major'.  */
#ifdef MAJOR_IN_MKDEV
#include <sys/mkdev.h>
#define HAVE_MAJOR
#endif
#ifdef MAJOR_IN_SYSMACROS
#include <sys/sysmacros.h>
#define HAVE_MAJOR
#endif
#ifdef major			/* Might be defined in sys/types.h.  */
#define HAVE_MAJOR
#endif

#ifndef HAVE_MAJOR
#define major(dev)  (((dev) >> 8) & 0xff)
#define minor(dev)  ((dev) & 0xff)
#define makedev(maj, min)  (((maj) << 8) | (min))
#endif
#undef HAVE_MAJOR

#ifdef HAVE_UTIME_H
#include <utime.h>
#endif

/* Some systems (even some that do have <utime.h>) don't declare this
   structure anywhere.  */
#ifndef HAVE_STRUCT_UTIMBUF
struct utimbuf
{
  long actime;
  long modtime;
};
#endif

/* Don't use bcopy!  Use memmove if source and destination may overlap,
   memcpy otherwise.  */

#ifdef HAVE_STRING_H
# if !STDC_HEADERS && HAVE_MEMORY_H
#  include <memory.h>
# endif
# include <string.h>
#else
# include <strings.h>
char *memchr ();
#endif

#include <errno.h>
#ifndef errno
extern int errno;
#endif

#ifdef STDC_HEADERS
#define getopt system_getopt
#include <stdlib.h>
#undef getopt
#else /* not STDC_HEADERS */
char *getenv ();
#endif /* STDC_HEADERS */

/* The following test is to work around the gross typo in
   systems like Sony NEWS-OS Release 4.0C, whereby EXIT_FAILURE
   is defined to 0, not 1.  */
#if !EXIT_FAILURE
# undef EXIT_FAILURE
# define EXIT_FAILURE 1
#endif

#ifndef EXIT_SUCCESS
# define EXIT_SUCCESS 0
#endif

#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#else
#include <sys/file.h>
#endif

#ifndef SEEK_SET
#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2
#endif
#ifndef F_OK
#define F_OK 0
#define X_OK 1
#define W_OK 2
#define R_OK 4
#endif

#ifdef HAVE_DIRENT_H
# include <dirent.h>
# define NLENGTH(direct) (strlen((direct)->d_name))
#else /* not HAVE_DIRENT_H */
# define dirent direct
# define NLENGTH(direct) ((direct)->d_namlen)
# ifdef HAVE_SYS_NDIR_H
#  include <sys/ndir.h>
# endif /* HAVE_SYS_NDIR_H */
# ifdef HAVE_SYS_DIR_H
#  include <sys/dir.h>
# endif /* HAVE_SYS_DIR_H */
# ifdef HAVE_NDIR_H
#  include <ndir.h>
# endif /* HAVE_NDIR_H */
#endif /* HAVE_DIRENT_H */

#ifdef CLOSEDIR_VOID
/* Fake a return value. */
#define CLOSEDIR(d) (closedir (d), 0)
#else
#define CLOSEDIR(d) closedir (d)
#endif

/* Get or fake the disk device blocksize.
   Usually defined by sys/param.h (if at all).  */
#ifndef DEV_BSIZE
#ifdef BSIZE
#define DEV_BSIZE BSIZE
#else /* !BSIZE */
#define DEV_BSIZE 4096
#endif /* !BSIZE */
#endif /* !DEV_BSIZE */

/* Extract or fake data from a `struct stat'.
   ST_BLKSIZE: Optimal I/O blocksize for the file, in bytes.
   ST_NBLOCKS: Number of 512-byte blocks in the file
   (including indirect blocks). */
#ifndef HAVE_ST_BLOCKS
# define ST_BLKSIZE(statbuf) DEV_BSIZE
# if defined(_POSIX_SOURCE) || !defined(BSIZE) /* fileblocks.c uses BSIZE.  */
#  define ST_NBLOCKS(statbuf) (((statbuf).st_size + 512 - 1) / 512)
# else /* !_POSIX_SOURCE && BSIZE */
#  define ST_NBLOCKS(statbuf) (st_blocks ((statbuf).st_size))
# endif /* !_POSIX_SOURCE && BSIZE */
#else /* HAVE_ST_BLOCKS */
/* Some systems, like Sequents, return st_blksize of 0 on pipes. */
# define ST_BLKSIZE(statbuf) ((statbuf).st_blksize > 0 \
			       ? (statbuf).st_blksize : DEV_BSIZE)
# if defined(hpux) || defined(__hpux__) || defined(__hpux)
/* HP-UX counts st_blocks in 1024-byte units.
   This loses when mixing HP-UX and BSD filesystems with NFS.  */
#  define ST_NBLOCKS(statbuf) ((statbuf).st_blocks * 2)
# else /* !hpux */
#  if defined(_AIX) && defined(_I386)
/* AIX PS/2 counts st_blocks in 4K units.  */
#   define ST_NBLOCKS(statbuf) ((statbuf).st_blocks * 8)
#  else /* not AIX PS/2 */
#   if defined(_CRAY)
#    define ST_NBLOCKS(statbuf) ((statbuf).st_blocks * ST_BLKSIZE(statbuf)/512)
#   else /* not AIX PS/2 nor CRAY */
#    define ST_NBLOCKS(statbuf) ((statbuf).st_blocks)
#   endif /* not _CRAY */
#  endif /* not AIX PS/2 */
# endif /* !hpux */
#endif /* HAVE_ST_BLOCKS */

/* Convert B 512-byte blocks to kilobytes if K is nonzero,
   otherwise return it unchanged. */
#define convert_blocks(b, k) ((k) ? ((b) + 1) / 2 : (b))

#ifndef RETSIGTYPE
#define RETSIGTYPE void
#endif

#ifdef __GNUC__
# undef alloca
# define alloca __builtin_alloca
#else
# ifdef HAVE_ALLOCA_H
#  include <alloca.h>
# else
#  ifndef _AIX
/* AIX alloca decl has to be the first thing in the file, bletch! */
char *alloca ();
#  endif
# endif
#endif

#include <ctype.h>

/* Jim Meyering writes:

   "... Some ctype macros are valid only for character codes that
   isascii says are ASCII (SGI's IRIX-4.0.5 is one such system --when
   using /bin/cc or gcc but without giving an ansi option).  So, all
   ctype uses should be through macros like ISPRINT...  If
   STDC_HEADERS is defined, then autoconf has verified that the ctype
   macros don't need to be guarded with references to isascii. ...
   Defining isascii to 1 should let any compiler worth its salt
   eliminate the && through constant folding."

   Bruno Haible adds:

   "... Furthermore, isupper(c) etc. have an undefined result if c is
   outside the range -1 <= c <= 255. One is tempted to write isupper(c)
   with c being of type `char', but this is wrong if c is an 8-bit
   character >= 128 which gets sign-extended to a negative value.
   The macro ISUPPER protects against this as well."  */

#if defined (STDC_HEADERS) || (!defined (isascii) && !defined (HAVE_ISASCII))
#define IN_CTYPE_DOMAIN(c) 1
#else
#define IN_CTYPE_DOMAIN(c) isascii(c)
#endif

#ifdef isblank
#define ISBLANK(c) (IN_CTYPE_DOMAIN (c) && isblank (c))
#else
#define ISBLANK(c) ((c) == ' ' || (c) == '\t')
#endif
#ifdef isgraph
#define ISGRAPH(c) (IN_CTYPE_DOMAIN (c) && isgraph (c))
#else
#define ISGRAPH(c) (IN_CTYPE_DOMAIN (c) && isprint (c) && !isspace (c))
#endif

#define ISPRINT(c) (IN_CTYPE_DOMAIN (c) && isprint (c))
#define ISALNUM(c) (IN_CTYPE_DOMAIN (c) && isalnum (c))
#define ISALPHA(c) (IN_CTYPE_DOMAIN (c) && isalpha (c))
#define ISCNTRL(c) (IN_CTYPE_DOMAIN (c) && iscntrl (c))
#define ISLOWER(c) (IN_CTYPE_DOMAIN (c) && islower (c))
#define ISPUNCT(c) (IN_CTYPE_DOMAIN (c) && ispunct (c))
#define ISSPACE(c) (IN_CTYPE_DOMAIN (c) && isspace (c))
#define ISUPPER(c) (IN_CTYPE_DOMAIN (c) && isupper (c))
#define ISXDIGIT(c) (IN_CTYPE_DOMAIN (c) && isxdigit (c))
#define ISDIGIT_LOCALE(c) (IN_CTYPE_DOMAIN (c) && isdigit (c))

/* ISDIGIT differs from ISDIGIT_LOCALE, as follows:
   - Its arg may be any int or unsigned int; it need not be an unsigned char.
   - It's guaranteed to evaluate its argument exactly once.
   - It's typically faster.
   Posix 1003.2-1992 section 2.5.2.1 page 50 lines 1556-1558 says that
   only '0' through '9' are digits.  Prefer ISDIGIT to ISDIGIT_LOCALE unless
   it's important to use the locale's definition of `digit' even when the
   host does not conform to Posix.  */
#define ISDIGIT(c) ((unsigned) (c) - '0' <= 9)

#ifndef __P
#if defined (__GNUC__) || (defined (__STDC__) && __STDC__)
#define __P(args) args
#else
#define __P(args) ()
#endif  /* GCC.  */
#endif  /* Not __P.  */

/* Take care of NLS matters.  */

#if HAVE_LOCALE_H
# include <locale.h>
#endif
#if !HAVE_SETLOCALE
# define setlocale(Category, Locale) /* empty */
#endif

#if ENABLE_NLS
# include <libintl.h>
# define _(Text) gettext (Text)
#else
# undef bindtextdomain
# define bindtextdomain(Domain, Directory) /* empty */
# undef textdomain
# define textdomain(Domain) /* empty */
# define _(Text) Text
#endif

#define STREQ(a,b) (strcmp((a), (b)) == 0)


src/oauthsign.c
/* 
   oauthsign - command line oauth

   Copyright (C) 2008, 2009 Robin Gareus

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

#include <termios.h>
#include <grp.h>
#include <pwd.h>
*/

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdio.h>
#include <sys/types.h>
#include <getopt.h>
#include "system.h"

#include "oauth_common.h"

#define EXIT_FAILURE 1

char *xmalloc ();
char *xrealloc ();
char *xstrdup ();

static void usage (int status);

/* The name the program was run with, stripped of any leading path. */
char *program_name;

/* getopt_long return codes */
enum {DUMMY_CODE=129
     ,DRYRUN_CODE
     ,NOWARN_CODE
     ,CURLOUT_CODE
     ,NULLCK_CODE
     ,NULLCS_CODE
     ,NULLTK_CODE
     ,NULLTS_CODE
};

/* Option flags and variables */

int no_warnings  = 0; /* --no-warn */
int want_quiet   = 0; /* --quiet, --silent */
int want_verbose = 0; /* --verbose */
int want_dry_run = 0; /* --dry-run */

int mode         = 1; ///< mode: 1=GET 2=POST; general operation-mode - bit coded 
                      //  bit0 (1)  : enable ?! (also see want_dry_run)
                      //  bit1 (2)  : HTTP POST enable (no GET) 
                      //  bit2 (4)  : (unused ; old oauthrequest() compat)
                      //  bit3 (8)  : -b base-string and exit
                      //  bit4 (16) : -B base-url and exit
                      //  bit5 (32) :  print secrets along with base-string(8)
                      //  bit6 (64) :  print signature after generating it. (unless want_verbose is set: it's printed anyway)
                      //  bit7 (128):  parse reply (request token, access token)
                      //  bit8 (256):  toggle ouput Parameter escape. (dont escape GETs and escape POST params with format_array(..)
                      //  bit9 (512):  curl-output
                      //
char *method;
int request_mode = 0; ///< mode: 0=print info only; 1:perform HTTP request

int print_as_get = 0; 
int want_write   = 0;
int   oauth_argc = 0;
char **oauth_argv = NULL;
char *datafile   = NULL;
char *callback = NULL;
char *verifier = NULL;
oauthparam op;

static struct option const long_options[] =
{
  {"quiet", no_argument, 0, 'q'},
  {"silent", no_argument, 0, 'q'},
  {"verbose", no_argument, 0, 'v'},
  {"dry-run", no_argument, 0, DRYRUN_CODE},
  {"no-warn", no_argument, 0, NOWARN_CODE},
  {"help", no_argument, 0, 'h'},
  {"version", no_argument, 0, 'V'},

  {"consumer-key", required_argument, 0, 'c'},
  {"consumer-secret", required_argument, 0, 'C'},
  {"token-key", required_argument, 0, 't'},
  {"token-secret", required_argument, 0, 'T'},
  {"CK", required_argument, 0, 'c'},
  {"CS", required_argument, 0, 'C'},
  {"TK", required_argument, 0, 't'},
  {"TS", required_argument, 0, 'T'},
  {"callback", required_argument, 0, 'a'},
  {"verifier", required_argument, 0, 'A'},
  {"erase-consumer-key", no_argument, 0, NULLCK_CODE},
  {"erase-consumer-secret", no_argument, 0, NULLCS_CODE},
  {"erase-token-key", no_argument, 0, NULLTK_CODE},
  {"erase-token-secret", no_argument, 0, NULLTS_CODE},
  {"signature-method", required_argument, 0, 'm'}, //  oauth signature method

  {"request", required_argument, 0, 'r'}, // HTTP request method (GET, POST)
  {"post", no_argument, 0, 'p'},
//{"toggle-escape", no_argument, 0, 'P'}, // TODO: rename ''--print-escaped'' ?!
  {"curl", no_argument, 0, CURLOUT_CODE}, 
  {"data", required_argument, 0, 'd'},
  {"base-url", no_argument, 0, 'B'},
  {"base-string", no_argument, 0, 'b'},

  {"file", required_argument, 0, 'f'},
  {"write", no_argument, 0, 'w'}, 
//{"writefile", required_argument, 0, 'F'}, 
//{"execute", no_argument, 0, 'x'}, 
//{"oauthrequest", no_argument, 0, 'X'}, 
  {NULL, 0, NULL, 0}
};

/** Set all the option flags according to the switches specified.
 *  Return the index of the first non-option argument.  
 */
static int decode_switches (int argc, char **argv) {
  int c;

  while ((c = getopt_long (argc, argv, 
			   "h"	/* help */
			   "V" 	/* version */
			   "q"	/* quiet or silent */
			   "v"	/* verbose */
			   "b"  /* base-string*/
			   "B"  /* base-URL*/
			   "r:" /* HTTP Request method */
			   "p" 	/* post */
			   "d:" /* URL-query parameter data eg.
               * '-d name=daniel -d skill=lousy'->'name=daniel&skill=lousy' */
			   "m:" /* oauth signature Method */
			   "P" 	/* print escaped post parameters */
			   "G" 	/* always print as GET parameters */

			   "c:" /* consumer-key*/
			   "C:" /* consumer-secret */
			   "t:" /* token-key*/
			   "T:" /* token-secret */

			   "f:" /* read key/data file */
			   "w" 	/* write to key/data file, save request/access token state */
			   "F:" /* set key/data filename */
			   "x" 	/* execute */
			   "X" 	/* execute and parse reply */
			   "e"  /* erase token */
			   "E", /* erase token and conusmer */
			   long_options, (int *) 0)) != EOF) {
    switch (c) {
      case 'q':		/* --quiet, --silent */
        want_quiet = 1;
        break;
      case 'v':		/* --verbose */
        want_verbose = 1;
        break;
      case NOWARN_CODE:	/* --no-warn */
        no_warnings = 1;
        break;
      case DRYRUN_CODE:	/* --dry-run */
        want_dry_run = 1;
        break;
      case CURLOUT_CODE:	/* --curl */
        mode|= 512;
        break;
      case 'V':
        printf ("%s %s (%s)", PACKAGE, VERSION, OS);
        #ifdef LIBOAUTH_VERSION
        printf (" liboauth/%s", LIBOAUTH_VERSION);
        #endif
        printf ("\n");
        exit (0);

      case 'b':
        mode&=~(8|16);
        mode|=8;
        break;
      case 'B':
        mode&=~(8|16);
        mode|=16;
        break;
      case 'p':
        mode&=~(1|2);
        mode|=2;
        free(method); method=xstrdup("POST");
        break;
      case 'r':
        mode&=~(1|2|4);
        if (!strncasecmp(optarg,"GET",3)) {
          mode|=1;
          free(method); method=xstrdup("GET");
     // else if (!strncasecmp(optarg,"POSTREQUEST",4))
     //   mode|=4;
        } else if (!strncasecmp(optarg,"POST",4)) {
          mode|=2;
          free(method); method=xstrdup("POST");
        } else {
          int i;
          mode|=1; // XXX
          method = strdup(optarg); 
          for (i=0;i<strlen(method);i++) method[i]=toupper(method[i]);
        }
        break;
      case 't':
        if (op.t_key) free(op.t_key);
        op.t_key=xstrdup(optarg); 
        break;
      case 'T':
        if (op.t_secret) free(op.t_secret);
        op.t_secret=xstrdup(optarg); 
        break;
      case 'a':
#if 1
        if (strlen(optarg) == 0) 
          callback=xstrdup("oob");
        else
#endif
        callback=xstrdup(optarg); 
        break;
      case 'A':
        verifier=xstrdup(optarg); 
        break;
      case 'c':
        if (op.c_key) free(op.c_key);
        op.c_key=xstrdup(optarg); 
        break;
      case 'C':
        if (op.c_secret) free(op.c_secret);
        op.c_secret=xstrdup(optarg); 
        break;
      case 'd': 
        add_param_to_array(&oauth_argc, &oauth_argv,optarg);
        break;
      case 'f':
        read_keyfile(optarg, &op);
      case 'F':
        if (datafile) free(datafile);
        datafile=xstrdup(optarg);
        break;
      case 'w':
        want_write=1;
        break;
      case 'X':
        mode|=128; // parse reply and enter request mode..
      case 'x':
        request_mode=1;
        break;
      case 'e':
        reset_oauth_token(&op);
        break;
      case 'E':
        reset_oauth_param(&op);
        break;
      case NULLCK_CODE:	/* --erase-consumer-key */
        if (op.c_key) free(op.c_key);
        op.c_key=NULL;
        break;
      case NULLCS_CODE:	/* --erase-consumer-secret */
        if (op.c_secret) free(op.c_secret);
        op.c_secret=NULL;
        break;
      case NULLTK_CODE:	/* --erase-token-key */
        if (op.t_key) free(op.t_key);
        op.t_key=NULL;
        break;
      case NULLTS_CODE:	/* --erase-token-secret */
        if (op.t_secret) free(op.t_secret);
        op.t_secret=NULL;
        break;
      case 'G':
        print_as_get=1;
        break;
      case 'P':
        mode|=256;
        break;
      case 'm':
        if (parse_oauth_method(&op, optarg)) usage(1);
        break;
      case 'h':
        usage (0);

      default:
        usage (EXIT_FAILURE);
    }
  }
  return optind;
}


static void
usage (int status)
{
  printf (_("%s - \
command line utilities for oauth\n"), program_name);
  printf (_("Usage: %s [OPTION]... URL [CKey] [CSec] [TKey] [Tsec]\n"), program_name);
  printf (_("\
Options:\n\
  -h, --help                  display this help and exit\n\
  -V, --version               output version information and exit\n\
  -q, --quiet, --silent       inhibit usual output\n\
  -v, --verbose               print more information\n\
  --no-warn                   dont print any warnings.\n\
  \n\
  -b, --base-string           print OAuth base-string and exit\n\
  -B, --base-url              print OAuth base-URL and exit\n\
"/*
  --curl                      format output as `curl` commandline\n\
*/"\
  -r, --request <type>        HTTP request type (HEAD, POST, GET [default],..)\n\
  -p, --post                  same as -r POST\n\
  -d, --data <key>[=<val>]    add ASCII url-query parameters.\n\
                              Note: oauthsign will URL-encode key & value.\n\
  -m, --signature-method <m>  oauth signature method (PLAINTEXT,\n\
                              RSA-SHA1, HMAC-SHA1 [default])\n\
"/*
  -P,                         print URL-escaped POST parameters\n\ 
*/"\
  \n\
  -c, --CK, --consumer-key    <text> \n\
  -C, --CS, --consumer-secret <text> \n\
  -t, --TK, --token-key       <text> \n\
  -T, --TS, --token-secret    <text> \n\
  \n\
  -a, --callback <url>        specify oauth_callback url (or \'oob') // 1.0 Rev A\n\
  -A, --verifier <text>       specify oauth_verifier // 1.0 Rev A\n\
  \n\
  -f, --file <filename>       read tokens and secrets from config-file\n\
  -w                          write tokens to config-file\n\
  -F <filename>               set config-file name w/o reading the file.\n\
  -x                          make HTTP request and return the replied content\n\
  -X                          make HTTP request and parse the reply for tokens\n\
                              use '-X -w' to request and store tokens.\n\
  --dry-run                   take no real actions (with -x, -w or -X)\n\
  -e, --erase-tokens          clear [access|request] tokens.\n\
  -E, --erase-all             wipe all tokens and reset method to HMAC-SHA1.\n\
  --erase-consumer-key        unset consumer-key\n\
  --erase-consumer-secret     unset consumer-secret\n\
  --erase-token-key           unset token-key\n\
  --erase-token-secret        unset token-secret\n\
  \n\
  The position of parameters -d, -f, -F, -e, -E and all tokens matters!\n\
  \n\
  Tokens are read from file at the moment the -f option is parsed overriding \n\
  the current value(s). Optional trailing key/secret params are parsed last.\n\
  eg.\n\
    '-f config.txt -e -C secret -F out.txt -w' reads the settings from file,\n\
  then deletes the access/request tokens and finally overrides the consumer-\n\
  secret. Only the consumer-key is left from config.txt and will be saved \n\
  to out.txt along with the new secret. If -X is given and the HTTP request\n\
  succeeds, the received token and secret will be stored as well.\n\
  \n\
  The request URL is constructed by first parsing all query-parameters from\n\
  the URL; then -d parameters are added, and finally oauth_XYZ params \n\
  appended.\n\
  \n\
"));
  exit (status);
}


int main (int argc, char **argv) {
  int i;
  int exitval=0;
  int oaargc =0;
  char **oaargv= NULL;
  char *sign = NULL;

  // initialize 

  program_name = argv[0];
  memset(&op,0,sizeof(oauthparam));
  reset_oauth_param(&op);
  method=xstrdup("GET");

  // parse command line

  i = decode_switches (argc, argv);

  if (i>=argc) usage(1);
  op.url=xstrdup(argv[i++]);

  if (argc>i) { if (op.c_key) free(op.c_key); op.c_key=xstrdup(argv[i++]); }
  if (argc>i) { if (op.c_secret) free(op.c_secret); op.c_secret=xstrdup(argv[i++]); }
  if (argc>i) { if (op.t_key) free(op.c_key); op.t_key=xstrdup(argv[i++]); }
  if (argc>i) { if (op.t_secret) free(op.t_secret); op.t_secret=xstrdup(argv[i++]); }
  if (argc>i) 
      usage(EXIT_FAILURE);

  // check settings 

  if (!op.c_key || strlen(op.c_key)<1) {
    fprintf(stderr, "ERROR: consumer key not set.\n");
    exit(1);
  }

  if (want_write && !datafile || (datafile && strlen(datafile)<1)) {
    want_write=0;
    fprintf(stderr, "ERROR: no filename given. use -F or -f.\n");
    exit(1);
  }

  if (!(mode&2) && print_as_get) {
    if (!no_warnings)
      fprintf(stderr, "WARNING: -G is redundant with GET requests.\n");
  } else if (!want_quiet && print_as_get) {
    if (!no_warnings)
      fprintf(stderr, "WARNING: non standard output format.\n");
  }
  if ((mode&256) && !want_quiet) {
    if (!no_warnings)
      fprintf(stderr, "WARNING: non standard parameter escape settings.\n");
  }

  if (want_write && !want_dry_run) { // save current state
    if (save_keyfile(datafile, &op)) {
      want_write=0;
      if (!no_warnings)
        fprintf(stderr, "WARNING: saving state to file '%s' failed.\n", datafile);
    } else if (want_verbose) {
      fprintf(stderr, "saved state to %s\n", datafile);
    }
  }

  if (callback) {
    add_kv_to_array(&oauth_argc, &oauth_argv, "oauth_callback", callback);
  }
 
  if (verifier) { // OAuth 1.0 Rev A
    add_kv_to_array(&oauth_argc, &oauth_argv, "oauth_verifier", verifier);
    free(verifier);
  }
  
  sign = oauthsign_ext(mode, method, &op, oauth_argc, oauth_argv, &oaargc, &oaargv);

  if (sign && want_verbose) 
    fprintf(stderr, "oauth_signature=%s\n", sign);

  if(!request_mode) {
    if (sign) { 
      add_kv_to_array(&oaargc, &oaargv, "oauth_signature", sign);
      free(sign);
    }
    if (print_as_get) mode&=~2; // print as GET !
    format_array(mode, oaargc, oaargv);
  } else { // request_mode 
    char *reply;
    if (!sign) { 
      if (!no_warnings && !want_quiet) 
        fprintf(stderr,"WARNING: could not generate oAuth signature.\n");
      exitval|=8;
    }

    if (!want_dry_run) {
      reply = oauthrequest_ext(mode, &op, oaargc, oaargv, sign);
    } else { 
      if (!want_quiet || want_verbose) 
        fprintf(stderr, "DRY-RUN. not making any HTTP request.\n"); 
      if (sign) add_kv_to_array(&oaargc, &oaargv, "oauth_signature", sign);
      if (print_as_get) mode&=~2; // print as GET !
      format_array(mode, oaargc, oaargv);
      reply=NULL;
    }

    if (!reply) { 
      if (!exitval || want_verbose) fprintf(stderr,"ERROR: no reply from HTTP request.\n");
      exitval|=2;
    } else if (want_verbose || (mode&128)==0) {
      if(want_verbose) fprintf(stderr, "------HTTP reply------\n");
      fprintf(want_verbose?stderr:stdout, "%s\n", reply);
      if(want_verbose) fprintf(stderr, "----------------------\n");
    }

    if (!want_dry_run && (mode&128)) {
      reset_oauth_token(&op);
      int reply_flags;
      if (parse_reply(reply, &(op.t_key), &(op.t_secret), &reply_flags)) { 
        if (!exitval || want_verbose) fprintf(stderr,"ERROR: could not parse reply.\n");
        exitval|=4;
      } else if (!want_quiet) {
        printf ("token=%s\n",op.t_key);
        printf ("token_secret=%s\n",op.t_secret);
        if((reply_flags&1) == 1 && callback) 
          printf ("oauth_callback_confirmed=true # -> OAuth V 1.0 Rev A\n");
      }
    }

    if (sign) free(sign);
    if (reply) free(reply);
  }

  if (callback) {
    // TODO: check if SP returned "oauth_callback_confirmed=true"  -> 1.0 rev A
    free(callback);
  }
 
  if (exitval==0 && want_write && !want_dry_run) { // save final state
    if (save_keyfile(datafile, &op)) {
      want_write=0;
      if (!no_warnings)
        fprintf(stderr, "WARNING: saving state to file '%s' failed.\n", datafile);
    } else if (want_verbose) {
      fprintf(stderr, "saved state to %s\n", datafile);
    }
  }
 
  free_array(oaargc, oaargv);
  free_array(oauth_argc, oauth_argv);
  free(method);
  return (exitval);
}

/* vim: set sw=2 ts=2 sts=2 et : */
